compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -2750,32 +2751,68 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
+                    let mut sugg = format!("{}, ", name.unwrap_or("'a"));
+
+                    if span.is_empty() {
+                        let mut generic_params = "".to_string();
+                        if let Some(with_poly_trait_ref) = self.with_poly_trait_ref.get(&node_id)
+                            && higher_ranked
+                        {
+                            generic_params = with_poly_trait_ref.generic_param_idents.iter().fold(
+                                "".to_string(),
+                                |mut generic_params, x| {
+                                    generic_params += x.as_str();
+                                    generic_params += ", ";
+                                    generic_params
+                                },
+                            );
+                            if !generic_params.is_empty() {
+                                rm_poly_trait_span = with_poly_trait_ref
+                                    .poly_trait
+                                    .with_hi(with_poly_trait_ref.trait_ref.lo());
+                            }
+                        }
+
+                        sugg = format!(
                             "{}<{}>{}",
                             if higher_ranked { "for" } else { "" },
-                            name.unwrap_or("'a"),
+                            format!("{}{}", generic_params, name.unwrap_or("'a")),
                             if higher_ranked { " " } else { "" },
                         );
-                        (span, sugg)
                     } else {
-                        let span = self
+                        span = self
                             .r
                             .tcx
                             .sess
                             .source_map()
                             .span_through_char(span, '<')
                             .shrink_to_hi();
-                        let sugg = format!("{}, ", name.unwrap_or("'a"));
-                        (span, sugg)
-                    };
+                    }
+
                     if higher_ranked {
                         let message = Cow::from(format!(
                             "consider making the {} lifetime-generic with a new `{}` lifetime",
                             kind.descr(),
                             name.unwrap_or("'a"),
                         ));
-                        should_continue = suggest(err, true, span, message, sugg);
+                        should_continue = if !rm_poly_trait_span.is_dummy() {
+                            // For poly-trait-ref like `for<'a> Trait<T>` in
+                            // `T: for<'a> Trait<T> + 'b { }`.
+                            // We should merge the higher-ranked lifetimes: existed `for<'a>` and suggestion `for<'b>`
+                            // or will get err:
+                            // `[E0316] nested quantification of lifetimes`.
```

Comment:
Ah I see now. Because 

```
T: for<'b> Trait<T> + 'b
```

doesn't make sense, as the `+ 'b` is not part of the `for<'b> Trait<T>`

Please document what is being removed and what is being added.

Reply:
Done. Thank you.